src/relay/op/tensor/nn.cc

```diff
@@ -0,0 +1,96 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file nn.cc
+ * \brief Property def of nn operators.
+ */
+#include <tvm/relay/expr.h>
+#include <tvm/relay/op.h>
+#include <tvm/relay/attrs/nn.h>
+#include "../type_relations.h"
+
+namespace tvm {
+namespace relay {
+
+TVM_REGISTER_NODE_TYPE(DenseAttrs);
+
+bool DenseRel(const Array<Type>& types,
+              int num_inputs,
+              const Attrs& attrs,
+              const TypeReporter& reporter) {
+  CHECK_EQ(types.size(), 3);
+  const auto* data = types[0].as<TensorTypeNode>();
+  if (data == nullptr) return false;
+  const auto* weight = types[1].as<TensorTypeNode>();
+  if (weight == nullptr) return false;
+
+  const DenseAttrs* param = attrs.as<DenseAttrs>();
+  CHECK(param != nullptr);
+
+  CHECK(static_cast<int>(data->shape.size()) != 0);
+  Array<tvm::Expr> oshape = data->shape;
+  oshape.Set((oshape.size() - 1), make_const(Int(64), param->units));
+
+  // assign output type
+  reporter->Assign(types[2],
+                   TensorTypeNode::make(oshape,
+                                        data->dtype));
+  return true;
+}
+
+// Positional relay function to create dense operator used by frontend FFI.
+Expr MakeDense(Expr data,
+               Expr weight,
+               int units) {
+  auto attrs = make_node<DenseAttrs>();
+  attrs->units = units;
+  static const Op& op = Op::Get("nn.dense");
+  return CallNode::make(op, {data, weight}, Attrs(attrs), {});
+}
+
+TVM_REGISTER_API("relay.op.nn._make.dense")
+.set_body([](const TVMArgs& args, TVMRetValue* rv) {
+    runtime::detail::unpack_call<Expr, 3>(MakeDense, args, rv);
+  });
+
+RELAY_REGISTER_OP("nn.dense")
```

Comment:
maybe add bias support?

Reply:
I think like conv where we kept bias away, we could do same here.